src/ledger/LedgerManagerImpl.cpp

```diff
@@ -662,12 +677,23 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
 
     ledgerClosed(ltx);
 
+    if (ledgerData.getExpectedHash() &&
+        *ledgerData.getExpectedHash() != mLastClosedLedger.hash)
+    {
+        throw std::runtime_error("Local node's ledger corrupted during close");
+    }
+
     if (mMetaStream)
     {
-        releaseAssert(ledgerCloseMeta);
-        ledgerCloseMeta->v0().ledgerHeader = mLastClosedLedger;
-        mMetaStream->writeOne(*ledgerCloseMeta);
-        mMetaStream->flush();
+        releaseAssert(mNextMetaToEmit);
+        mNextMetaToEmit->v0().ledgerHeader = mLastClosedLedger;
+        // If the LedgerCloseData provided an expected hash, then we validated
+        // it above.
+        if (mApp.getConfig().UNSAFE_EMIT_META_EARLY ||
+            ledgerData.getExpectedHash())
+        {
+            emitNextMeta();
+        }
```

Comment:
I would change the code to set `mNextMetaToEmit` in the else here and use `ledgerCloseMeta` before

Reply:
(also fixed)